File: daily/2025/3363_Find_the_Maximum_Number_of_Fruits_Collected.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int maxCollectedFruits(vector<vector<int>>& fruits) {
        int n = fruits.size();
        int ans = 0;
        for (int i = 0; i < n; i++){
            ans += fruits[i][i];
            fruits[i][i] = 0;
        }
        vector<vector<int>> dp(n, vector<int>(n, 0));
        dp[0][n-1] = fruits[0][n-1];
        for (int i = 1; i < n; i++){
            dp[i][n-1] = fruits[i][n-1] + max(dp[i-1][n-1], dp[i-1][n-2]);
            for (int j = n - 2; j >= max(i, n - 1 - i); j--){
                dp[i][j] = fruits[i][j] + max({dp[i-1][j-1], dp[i-1][j], dp[i-1][j+1]});
            }
        }
        // for (auto i : dp){
        //     for (auto j : i) cout << j << " ";
        //     cout << endl;
        // }
        ans += dp[n-1][n-1];
        // dp = vector<vector<int>>(n, vector<int>(n, 0));
        dp[n-1][0] = fruits[n-1][0];
        for (int j = 1; j < n; j++){
            dp[n-1][j] = fruits[n-1][j] + max(dp[n-1][j-1], dp[n-2][j-1]);
            for (int i = n-2; i >= max(j, n-1-j); i--){
                dp[i][j] = fruits[i][j] + max({dp[i-1][j-1], dp[i][j-1], dp[i+1][j-1]});
            }
        }
        ans += dp[n-1][n-1];
        return ans;
    }
};

int main(){
    vector<vector<int>> fruits = {{1,2,3,4},{5,6,8,7},{9,10,11,12},{13,14,15,16}};
    // vector<vector<int>> fruits = {{1,1},{1,1}};
    Solution S;
    cout << S.maxCollectedFruits(fruits) << endl;
    return 0;
}

File: daily_python/3363_Find_the_Maximum_Number_of_Fruits_Collected.py
import os
import sys
from typing import List

class Solution:
    def maxCollectedFruits(self, fruits: List[List[int]]) -> int:
        n = len(fruits)
        ans = 0
        for i in range(n):
            ans += fruits[i][i]
        dp = [0] * n
        prev = [0] * n
        prev[n-1] = fruits[0][n-1]
        for i in range(1, n-1):
            for j in range(max(i+1, n-1-i), n):
                best = prev[j]
                if j > 0:
                    best = max(best, prev[j-1])
                if j < n-1:
                    best = max(best, prev[j+1])
                dp[j] = fruits[i][j] + best
            dp, prev = prev, dp
        ans += prev[n-1]
        dp = [0] * n
        prev = [0] * n
        prev[n-1] = fruits[n-1][0]
        for j in range(1, n-1):
            for i in range(max(j+1, n-1-j), n):
                best = prev[i]
                if i > 0:
                    best = max(best, prev[i-1])
                if i < n-1:
                    best = max(best, prev[i+1])
                dp[i] = fruits[i][j] + best
            dp, prev = prev, dp
        ans += prev[n-1]
        return ans

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            fruits = [list(map(int, row.split(','))) for row in lines[0].split('=')[1].strip()[2:-3].split('},{')]
            dataset.append(fruits)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for fruits in dataset:
        results.append(solution.maxCollectedFruits(fruits))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')